IP-423: Concurrent delivery

mezis opened the pull request:
This branch aims to implement #38.

~So far, it presents the future internal data model and execution flows.~

### What changes

The **API is entirely unchanged**; the integration tests are entirely unmodified (apart from a couple cosmetic changes)

Some of the behaviour does change:

- Batching will be favoured. The bus now fills up the batches up to the deadline instead of having a tendency to break batches down. In practice, with the defaults (batches of 100 and a deadline of 500ms), if you send more than 100 events per 500ms, you'll only get batches of 100.
- Delivery will back off exponentially when delivery fails. If your subscriber goes down (or times out), the next attempt will back off by 1, 2, ... 64 seconds (with some randomness added). 


### Why is this diff so big

The new internal design is now documented in [`INTERNALS.md`](https://github.com/deliveroo/routemaster/blob/3949b4d56b8ed2b29be2c03212beee187a2a6dbd/INTERNALS.md); it is recommended to read this first. 

The rearchitecture aims mainly to allow the delivery backend ("worker", formerly "watch") to scale out based on _throughput_, without the former limitations on number subscribers, as well as enable new features (e.g. delivery backoff).

Event delivery workers becomes subscriber-agnostic; they just deliver from a master queue of batches.
They also become multi-threaded, as is suitable for their I/O bound workload.
They're also just-in-time, as each delivery worker thread listens for batches using Redis's block-pop operations.

Batching is now performed on ingestion, atomically — in fact run as a single Lua script on the Redis server.

![](https://www.dropbox.com/s/j8q7yy2pbmzei3i/Screenshot%202016-10-27%2010.39.41.png?raw=1)

![](https://www.dropbox.com/s/7ai48fna0y64mp2/Screenshot%202016-10-27%2010.39.51.png?raw=1)

===

Jira story [#IP-423](https://deliveroo.atlassian.net/browse/IP-423) in project *Infrastructure and Platform*:

> (see also https://github.com/deliveroo/routemaster/issues/38)
> 
> ## Why
> 
> Event delivery (the _watch_ process) is currently single threaded, which
> - is inefficient (most of the time is spent waiting for delivery to complete)
> - cause timing issues (delivery backoff is hard to implement, as the watch waits for the earliest subscriber to deliver next between runs, and attempts delivery to all subscribers at each iteration).
> - scales poorly with the number of subscribers (each delivery process has a thread for each subscriber)
> 
> ## What
> 
> Change the data model, and delivery, to work of a global queue of _batches_ instead of per-subscriber queues of _events_.
> 
> https://github.com/deliveroo/routemaster/issues/38 has a discussion on an implementation approach.

mezis committed:
IP-423: WIP - data and execution model

levriero commented:
> will be even more important once we start accepting payloads with events

Have we decided to support this going forward? I thought it was something we wanted to avoid.

mezis commented:
> Have we decided to support this going forward? I thought it was something we wanted to avoid.

@gbonfant: yes, we'll support this - with very strict restrictions. It's still a Bad Idea, but there are edge cases (e.g. the `rider_location` resource) where the "pure" alternative is overly difficult to build. Compromise! See [IP-426](https://deliveroo.atlassian.net/browse/IP-426)

nronas commented:
LGTM 👍

gregbeech commented on INTERNALS.md:
```diff
@@ -49,49 +49,160 @@ This process runs scheduled tasks, regularly calling the following services:
-  - `timeout`: how long to defer event delivery for batching purposes.
+  - `timeout`: how long to defer event delivery for batching purposes (aka deadline).
   - `max_events`: maximum number of events to batch.
   - `uuid`: the credential to use when delivering events.
```
Should this `uuid` also be changed to `token` ?

gregbeech replied:
Ah no, ignore me, I see you're separating the two concepts.

gregbeech reviewed (approved):
:shipit:

mezis committed:
WIP - New batch model

mezis committed:
WIP - Ingestion/delivery cycle

mezis committed:
WIP - Bump Ruby

mezis committed:
WIP - Generic, diminutive job engine

mezis committed:
WIP - Cleanup refs to old code

mezis committed:
WIP - passing integration tests

mezis committed:
WIP - improved job promotion

mezis committed:
WIP - missing specs

mezis committed:
WIP - failure modes, Lua conversion

mezis committed:
WIP - fixes auto-dropper

mezis committed:
WIP - makes autodrop a job

mezis committed:
WIP - fixes slow ticker threads

mezis committed:
WIP - ported monitoring

mezis committed:
WIP - doc fixup

mezis committed:
WIP - fixes Travis config

mezis committed:
WIP - Code climate fixes

mezis committed:
WIP - wisper specs

mezis committed:
WIP - support concurrent event ingestion

mezis committed:
WIP - updated procfile

mezis committed:
WIP - better thread pool management

mezis committed:
WIP - Removes spurious comment

mezis committed:
Make Code Climate happy

mezis committed:
Code Climate! ❄️🌨🌩

mezis committed:
WIP - removes obsolete pseudocode

mezis committed:
Fixes Procfile

mezis committed:
WIP - fixes redis pool for ingestion

mezis committed:
WIP - promote full batches

mezis committed:
WIP - auto-promote batches on ingestion

mezis committed:
WIP - better ingestion concurrency

mezis committed:
WIP - fully deterministic ingestion

mezis committed:
Better docs

mezis committed:
Tweaks

sideshowcoder commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,245 @@
+        def _batch_ref_key(name)
+          name ? "batches:current:#{name}" : nil
+        end
+
```
Are the following keys global? I wondering about it for in case of sharding or is this not something we worry about?

mezis replied:
They're global. We're not sharding here (so far).
Some would merit sharding at some point though, e.g. the `batches:index` ZSET - so that we keep O(1) index ops instead of O(log n).

sideshowcoder replied:
I was more worried about the fact that with unshareded we are binding ourselves to 1 redis instance meaning everything is slow as soon as one operation is slow. In general do we keep sharding in mind here? spreading multiple queues over multiple redissss (what is the plural of redis 😁 ) seems reasonable. (there are more keys which seem global thats why I'm concerned)

sideshowcoder commented on routemaster/models/job.rb:
```diff
@@ -0,0 +1,45 @@
+
+      def initialize(name:, args:[], run_at:nil, data:nil)
+        @name   = name
+        @args   = args.kind_of?(Array) ? args : [args]
```
Same as `Array(args)` no need to check `kind_of?`

sideshowcoder commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,192 @@
+        attr_reader :delay
 
-      attr_reader :subscriber
+        DEFAULT_DELAY = 10_000 # milliseconds
```
Maybe better `DEFAULT_MAX_DELAY`

mezis replied:
Well, no. It's not a maximum — and in fact the delay can go up to 96,000.
But this is ugly though; I'll remove the default as a delay explicitely gets passed in all cases.

mezis committed:
Addresses PR comments

mezis committed:
Removes FIXMEs for already-logged work

mezis committed:
Merge branch 'master' into IP-423/concurrent-delivery

mezis commented:
Note to readers: the PR seems already approved, but it's the design that's been approved (As of writing), not the code.

gregbeech commented on .codeclimate.yml:
```diff
@@ -0,0 +1,29 @@
+      - ruby
+      - javascript
+      - python
+      - php
```
Being a bit picky, but do we need anything other than Ruby in here? Same for the rating paths below.

gregbeech commented on .ruby-version:
```diff
@@ -1 +1 @@
-2.3.1
+2.3.3
```
If we're feeling brave we could go to 2.4.0 now.

mezis replied:
There _are_ performance improvements in 2.4.0, true. If it doesn't require any code changes I'll bump :+1:

mezis replied:
... tried, and a number of gems (with no available upgrade) warn of `warning: constant ::Fixnum is deprecated` amongst other things.

It's probably harmless, but does make for a bigger change, so... rain check?

gregbeech replied:
Sure, the warnings are annoying, but most of them have open PRs so should be fixed soon. The logistics-dashboard is running 2.4.0 as of today, so we can use that as a testbed.

gregbeech commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+### Worker process
 
-This process is built from 4 key classes:
+This process (of which multiuple instances can be run) executes a number of
```
`s/multiuple/multiple`

gregbeech commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
-- The `Deliver` service gracefully sends event batches over HTTP;
-- The `Queue` model abstracts out queue management with Redis, providing a
-  syncronous means to push and pop message from a queue.
+- A group a `ROUTEMASTER_WORKER_THREADS` threads runs the `main` job queue
```
"A group _of_" ?

gregbeech commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
   A hash of subscription medatata. Keys:
   - `callback`: the URL to send events to.
-  - `timeout`: how long to defer event delivery for batching purposes.
+  - `timeout`: how long to defer event delivery for batching purposes (aka deadline).
```
I realise this isn't a change, but I'm not sure `timeout` is a very good name here as it would lead me to believe it's related to the receive timeout rather than the event staleness. Maybe `max_staleness` or something along those lines would be a more descriptive key?

mezis replied:
Totally agreed. There's work planned for this; my aim with this (large) change was to not change the API _at all_. Once merged, I'm planning to revisit/expand #46 to rename `timeout` as well (while preserving backwards compatibility).

gregbeech commented on core_ext/hash.rb:
```diff
@@ -6,4 +6,12 @@ def symbolize_keys
+        h[key] = yield h.delete(key)
+      end
+    end
+  end
```
If we move to Ruby 2.4 then this is a built-in function.

gregbeech commented on routemaster.rb:
```diff
@@ -2,4 +2,28 @@ module Routemaster
+  def self.config
+    {
+      redis_pool_size: 1
+    }.merge(@_config || {})
```
Would it be worth memoizing this so there isn't a hash allocation each time it's called? (Or alternatively doing the merge in the `configure` method?

gregbeech commented on routemaster/jobs/batch.rb:
```diff
@@ -0,0 +1,57 @@
+        if subscriber.nil? || data.nil? || data.empty?
+          batch.delete
+          return self
+        end
```
This feels like it should be a method on `Models::Batch`, e.g. `deliverable?` or something along those lines which would make this (rather long) method shorter, and more testable.

mezis replied:
Great point. This will make the abstraction less leaky.

gregbeech commented on routemaster/jobs/batch.rb:
```diff
@@ -0,0 +1,57 @@
+
+        events = data.
+          map { |d| Services::Codec.new.load(d) }.
+          select { |msg| msg.kind_of?(Models::Event) }
```
Similarly here, I'd possibly make this an `events` method on `Batch` as it seems that _it_ should be what knows its own structure rather than jobs/services (to be honest, I'd be inclined to hide the internal data structure of a batch as that's an implementation detail and only expose the actual interface that's desired by clients; i.e. nobody outside `Models::Batch` should really be using the raw data).

mezis replied:
Not sure I 100% agree here. At the moment, `Batch` is completely ignorant of the structure of messages (they're just raw blobs). The data structure of the batch itself is abstracted (which Redis keys, how they're indexed); but the structure of the _payloads_ is (currently, from the perspective of `Batch`) irrelevant.

I'm not inclined on adding `Event` serialisation/deserialisation into `Batch` itself as a) it's not necessary for `Batch` to work and b) it's already quite meaty.
Do you have an alternate suggestion? Slapping an abstraction layer atop `Batch`, maybe subclassing into `EventBatch`? Something else?

gregbeech replied:
I might be misunderstanding things... can anything other than an event be in a batch?

gregbeech commented on routemaster/mixins/redis.rb:
```diff
@@ -12,13 +11,10 @@ def self.included(by)
-      def _redis_prefix
-        @@_redis_prefix ||= ENV.fetch('ROUTEMASTER_REDIS_PREFIX')
+        @@_redis ||=
+        ConnectionPool.wrap(size: Routemaster.config[:redis_pool_size], timeout: 2) do
```
Should we really be using `wrap` here? Is there a plan to migrate to using `.with` rather than wrapping?

mezis replied:
I considered this; but I don't think it'd be that useful, because:

- most of the Redis usage is Lua scripts (there rarely is a sequence of calls that'd benefit from being made from the same connection.
- other places use `multi`, which is implicitly equivalent (it holds a lock on the connection). 
- the cost of the pool locking is minimal compared to the Redis roundtrip.

Let me know if that convinces you.

gregbeech replied:
My understanding is that `with` avoids dynamic method dispatch using `method_missing` rather than anything to do with the connections/locking/etc. The cost of the method dispatch is likely fairly insignificant in the grand scheme of things, but given it's not a hard change to make it might be worth doing.

gregbeech commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,244 @@
+      attr_reader :uid, :deadline
+
+
+      def initialize(uid:, deadline:nil, subscriber:nil)
```
Mildly picky, but `nil` should have a space before in both cases.

gregbeech commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,244 @@
+      def length
+        @_length ||= begin
+          raw = _redis.llen(_batch_key)
+          return if raw.nil? || raw.zero?
```
Why not `return 0` so callers don't have to handle `nil`? I saw `batch.length || 0` above and thought it was a bit strange. Now I see why...

mezis replied:
Well, that wouldn't be correct. A "broken" batch (no key, or no elements) doesn't have a length of zero. I could raise an exception instead if you think that's more readable?
(although the refactor you suggested with `Batch#deliverable?` might make this go away)

gregbeech replied:
If it's a 'broken' batch then yeah an exception sounds like it would make more sense.

gregbeech commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,244 @@
+
+
+        def scrub
+          raise NotImplementedError
```
?

mezis replied:
Place holder - batch scrubbing will be implemented in IP-283. Happy to remove it to keep this diff cleaner though.

gregbeech commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,244 @@
+          @batch_size = batch_size
+        end
+
+        # Yied all know batches, in creation order.
```
"Yield all known" ?

gregbeech commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,189 @@
+
+      # Returns a list of all queued jobs.
+      # Only use this for testing.
+      def dump
```
`dump` seems a weird name here; that's normally for serializing to an output format. `jobs` seems more like what it's actually doing?

gregbeech commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,189 @@
-        Services::Codec.new.load(payload, uid)
+      
+      # Returns the job currently being run by `worker_id`.
+      def running(worker_id)
```
Being picky, `running_jobs` might be a better name here as it appears to return an array of running jobs; the name `running` made me think it was whether the queue itself was running.

gregbeech commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,189 @@
+      # Returns the job currently being run by `worker_id`.
+      def running(worker_id)
+        raw = _redis.lrange(_pending_key(worker_id), 0, 0)
+        return unless raw
```
Maybe `return []` so callers don't have to handle `nil`?

gregbeech commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,189 @@
+        return if job_data.nil?
+
+        job = Job.load(job_data)
+        # _log.debug { "queue.#{@name} yiedling #{job.inspect}" }
```
Remove the commented out line?

gregbeech commented on spec/integration/delivery_spec.rb:
```diff
@@ -68,6 +68,7 @@ def subscribe
     5.times do |index|
       client.created('cats', "https://example.com/cats/#{index}")
     end
+    # binding.pry
```
Should probably delete this.

gregbeech commented on spec/jobs/autodrop_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+describe Routemaster::Jobs::Autodrop do
+  subject { described_class.new(batch_size: 2, database: database) }
+  let(:database) { double 'database' }
```
Any reason not to use `instance_double` here so that the tests ensure we're only using methods that actually exist?

mezis replied:
No reason beyond (bad) muscle memory. :+1:

gregbeech commented on spec/jobs/batch_spec.rb:
```diff
@@ -0,0 +1,98 @@
+require 'spec/support/env'
+require 'routemaster/jobs/batch'
+require 'routemaster/services/codec'
+# require 'routemaster/services/deliver'
```
Remove this line?

gregbeech commented on spec/jobs/batch_spec.rb:
```diff
@@ -0,0 +1,98 @@
+      end
+    }
+    let(:messages) { [make_event, make_event] }
+    let(:delivery) { double 'delivery', call:nil }
```
Ditto above comment about `instance_double`.

mezis replied:
Actually, this one is a stand-in for the `Services::Deliver` _class_, so `instance_double` wouldn't help? (it'd be an instance of `Class`?)

gregbeech replied:
I'm pretty sure if you did `instance_double(Services::Deliver, ...` then you'll get the class methods of that class. `instance_double` is badly named and doesn't really have anything to do with instances...

gregbeech commented:
I've done the best I can to review it and it all seems generally reasonable, but it's obviously a _huge_ amount of code and I'm sure there are things I've missed (inevitable with this much change). My suspicion is that we'll need to get it onto staging (and, really, production) and give it a good hammering to find the remaining issues.

mezis commented:
> My suspicion is that we'll need to get it onto staging ... and give it a good hammering

Definitely part of the plan, and I've started this. I've already used the [`routemaster-rushhour`](https://github.com/deliveroo/routemaster-rushhour#routemaster-rush-hour) stress/test harness with this branch, to check performance and scalability. Didn't miss a single event out of 10s of millions (I'll paste a link to the updated stress-test results).

mezis commented:
Here are some benchmark [results](https://github.com/deliveroo/routemaster-rushhour/blob/a842fef040e2ffaa6757b6442deb16053f03b006/RESULTS.md).

Notably: I tested the bus with an event throughput 3x higher than the total number of database writes we currently do on our monolith at peak (1000/s), i.e. what would happen at worst if we sent events for _every_ model change, and only loaded the backing store at 5%. Plenty of headroom.

mezis committed:
Addresses PR comments

mezis committed:
Addresses PR comments

mezis committed:
Happy climate ⛅️ + clarifying comment

mezis committed:
Addresses async spec flakiness

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
-
-- `Monitor` delivers metrics to metric adapters;
+- `Monitor` delivers gauge metrics to metric adapters;
 - `Autodrop` automatically removes the oldest messages from queues under low
```
This is the first time we are mentioning `Autodrop` in new version of document. The only services mentioned are `web`, `watch` and `monitor`.

mezis replied:
`web`, `watch`, and `monitor` are processes, not services (as in, service classes). I'll try to rephrase to clarify though!
(using the "job" terminology which is probably more familiar)

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+(high concurrency) more than one current batch might get created.
+
+When the batch is either full or stale, it gets promoted to _ready_ - another
+_current_ batch may then be created which the first awaits delivery.
```
which -> while

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+_current_ batch may then be created which the first awaits delivery.
+
+Finally, batches get deleted once they have been delivered; or when the
+auto-dropper removes them (because the system is running out of storage space).
```
Would we get some warning/alerts when the system starts doing this for a topic/subscriber?

mezis replied:
Yes: monitoring (which I've kept separate from this) includes auto-dropping in #50.

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
 `topic:{name}`
 
   A hash containing metadata has about a topic. Keys:
   - `publisher`: the UUID of the (singly authorized) publisher
```
Could you expand on term `singly authorized` here?

mezis replied:
I'll add a note. Simply put - only a single publisher is allowed to emit events for a given topic (this helps enforce single-service authority on resources/topics).

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+
+`jobs:scheduled:{q}` (sorted set)
+
+  Jobs scheduled for later execution, scored by their deadline timestamp.
```
may be add "in the queue named `{q}`" ?

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
+`jobs:queue:{q}` (list)
 
-  A list of UIDs of messages to be delivered, in reception order.
+  Jobs in the "instant" state, for immediate execution.
```
following the established nomenclature, shouldn't key here be more explicit. e.g `jobs:instant:{q}`

faheemmughal commented on INTERNALS.md:
```diff
@@ -23,75 +23,188 @@ Serves endpoints through 3 controllers:
-  timestamp of the attempt.
-  This gets cleared when messages are acked or nacked.
+  A list of at most 1 element, containing the job currently being processed by
+  worker `{worker}`.
```
why should it be a list?

mezis replied:
Because that allows us to `BRPOPLPUSH` (block-pop from the queue, push to the worker's "pending list"), which is interrupt-safe (ie. you'll never lose jobs).
This is what, for instance, Sidekiq Pro uses internally.

mezis replied:
(added a clarification, but below this comment this this didn't get hidden.)

faheemmughal commented on routemaster.rb:
```diff
@@ -2,4 +2,30 @@ module Routemaster
+
+  DEFAULTS = {
+    redis_pool_size: 1
+  }.freeze
```
interesting. Why .freeze?

mezis replied:
Because you wouldn't want to accidentally modify the defaults hash? It's not required. I can remove if it's confusing.

sideshowcoder replied:
doesn't the fact it's a constant already issue at least a warning on modification?

mezis replied:
Nope, unfortunately. It's the reference that's constant, not the value:

```ruby
> FOO = { a: 1 }
#=> {:a=>1}
> FOO[:b] = 2
#=> 2
# no warning
```

sideshowcoder replied:
BAD RUBY 🗡

mezis replied:
well... try `const u_int32*` in C. You get a constant pointer to a mutable integer :)
(or is it the other way around? mystery.)

mezis committed:
IP-423: Addresses PR comments (docs, naming)

sideshowcoder commented on routemaster/services/thread.rb:
```diff
@@ -0,0 +1,62 @@
+
+      def stop
+        _log.info { "stopping #{name}" }
+        @running = false
```
Shouldn't that call `Thread.stop` somewhere?

mezis replied:
`Thread#stop` would just interrupt the thread, without giving it a chance to clean up. It might, for instance, stop a delivery job after it delivers a batch but before it marks it as delivered (which would be fine as the data would get recovered, but would result in a redelivery)  

This pattern (with `@running`) simply signals that the thread should stop; the `wait` method below calls `Thread#join`, which waits for the thread to complete.

sideshowcoder replied:
mh ok got confused with the naming then, as I expected it to actually stop the thread not wait for it to finish whatever it is doing.

mezis replied:
Well, I like to think of method calls as messages — `stop` tells our thread wrapper object to stop, `wait` waits for it to finish whatever it's doing. It's not good if it's confusing though, so I'll take suggestions for renaming those methods?

(given they're almost actors, I could rename `Routemaster::Services::Thread(Group)` to `...Actor(Group)` but that might make things worse?)

sideshowcoder commented on bin/worker:
```diff
@@ -0,0 +1,52 @@
+
+  _log.info { 'trapping signals for clean exit' }
+  %w(INT TERM QUIT).each do |signal|
+    Kernel.trap(signal) { Thread.new { group.stop } }
```
Is it safe to call `stop` more than once on a group?

mezis replied:
It is: the method is idempotent. Like `Routemaster::Services::Thread#stop` above, it just signals the individual threads to stop and closes the error queue (itself idempotent), which stops the error watcher thread.

sideshowcoder commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,247 @@
+
+      # Is this batch deliverable?
+      def valid?
+        subscriber && data&.any?
```
what is `data&.any?`?

mezis replied:
Ruby 2.1 introduced the `&.` operator; this is conviently equivalent to (and slightly faster than)

```ruby
subscriber && (data && data.any?)
```

sideshowcoder replied:
Nice 👍

sideshowcoder commented on routemaster/models/batch.rb:
```diff
@@ -1,56 +1,247 @@
+        end
+
+        def _batch_key(uid)
+          uid ? "batch:#{uid}" : nil
```
Does that mean we have to deal with `_batch_key` being `nil` where it is used?

mezis replied:
We do. Is this a concern? (it is handled where relevant, I believe)

sideshowcoder replied:
I was wondering because there are cases where there is no `nil` check for the `_batch_key` as I have seen, and I wondered what the relevance of `nil` is here aka when is `nil` a valid `_batch_key` and maybe this case should get a name as a concept, leaking `nil` always makes me uneasy as catching all the cases seems hard.

sideshowcoder commented on routemaster/models/queue.rb:
```diff
@@ -1,130 +1,189 @@
-        return if uid.nil?
+      # Count all jobs; or all jobs before the deadline, if specified.
+      def length(deadline:nil, scheduled: true, instant: true)
+        deadline ||= '+inf'
```
why not just make it `'+inf'` in the signature instead of `nil`

mezis replied:
That would be a leaky abstraction? As the caller would "see" `+inf` which is a Redis specific implementation detail.
I could add an assertion; the intent is that the `deadline` argument is either `nil` or a timestamp.

sideshowcoder commented:
I didn't get through completely, yet it's just a lot, I think @gregbeech is right we should get it deployed and see it work.

About the performance, I only saw the median in the measurements, did you measure 95th percentile as well? Given the number of events we expect seeing the distribution seems interesting.

https://github.com/deliveroo/routemaster-rushhour/blob/a842fef040e2ffaa6757b6442deb16053f03b006/RESULTS.md#delivery-performance

mezis commented:
@sideshowcoder: replying to your comments here as some don't have a reply button! (weirdly)

> I was more worried about the fact that with unshareded we are binding ourselves to 1 redis instance meaning everything is slow as soon as one operation is slow. In general do we keep sharding in mind here? spreading multiple queues over multiple redissss (what is the plural of redis 😁 ) seems reasonable. (there are more keys which seem global thats why I'm concerned)

I'd say that's "Redises". Joke aside, after discussing with @nronas and @gregbeech we decided not to support sharding for now (high-availability replication aside, of course) because even under the most aggressive testing I could manage (spinning up 240 threads all sending events as fast as possible to the bus), I [didn't manage to load Redis over 5%](https://www.dropbox.com/s/1e7ateiyctjmzx1/Screenshot%202017-01-05%2018.07.00.png?dl=0). On a small development instance. And most of that is because the instance wasn't HA, thus writing to disk — otherwise it's 3%. Redis is really, really fast, especially given we're doing pretty much everything with small Lua scripts.

Given that, my hunch is that bus redundancy would be better achieved with a pool of monolithic busses (ie. spinning up multiple Routemasters each backed by a single Redis) with a client-side round-robin dispatch and/or circuit breaker (or a proxy).

> everything is slow as soon as one operation is slow

I've carefully been reviewing the Redis operations we perform here — you're most welcome to have a look too of course. 
Under load, the slowest queries are the Lua scripts (obviously), but those [peak at 25µs](https://www.dropbox.com/s/xa579efq1i11xwr/Screenshot%202017-01-05%2018.12.33.png?dl=0).

Most operations we perform are _O(1)_, a few are _O(log n)_ where _n_ is the number of queued jobs or batches. Also, the Redis instance is dedicated, and backups should be taken off a replica, so there isn't a lot of opportunity for any Redis query to be slow.

Even if a subscriber falls behind a lot, things stay fast — although I'll add specific benchmarks for this to convince you.

> About the performance, I only saw the median in the measurements, did you measure 95th percentile as well? Given the number of events we expect seeing the distribution seems interesting.

All the benchmarks I [posted](https://github.com/deliveroo/routemaster-rushhour/blob/a842fef040e2ffaa6757b6442deb16053f03b006/RESULTS.md#delivery-performance) do have 5th and 95th percentile measurements as well?
Empirically, end-to-end latency at scale is mostly driven by the ability of subscribers to keep up though. #50 and #51 introduce many more metrics, and while this is getting reviewed I'm prepping more benchmarking which will include Datadog dashboard screenshots. Let me know if there's specific things we'd like to measure.

> Given the number of events we expect seeing the distribution seems interesting.

As mentioned above, I'm benchmarking with 1,000 event/s in mind — which is 3x the volume of writes we _currently_ do on our master DB (so about what would happen if we sent events for every single model in the monolith, ie. an overestimate).

Do you expect many more events than that? (if so, I need to ramp up my benchmarking)

sideshowcoder commented:
Ah I was looking at the delivery performance median and missed the 95 in the at scale section, sorry about this. I don't think more events make sense, I do think so evaluating at the p95 is useful as I was looking at the delivery latencies (which are median) it came to my mind.

Redundancy by having multiple routemasters makes sense yes, and more metrics are highly welcome 😁

mezis committed:
IP-423: Addresses PR comment

mezis committed:
Merge branch 'master' into IP-423/concurrent-delivery

mezis committed:
IP-243: Fixes thread group cleanup

Queue closing should be done after the threads as stop - as threads
might concurrently fail.

mezis committed:
Merge branch 'master' into IP-423/concurrent-delivery

mezis commented:
This is on staging, and working. It's been pretty aggressively tested. Time to merge - any further objections, please feel free to nudge me and/or raise a ticket.
